etcdserver/cluster_util.go

```diff
@@ -177,6 +177,99 @@ func getVersions(lg *zap.Logger, cl *membership.RaftCluster, local types.ID, rt
 	return vers
 }
 
+// decideAllowedVersionRange decides the available version range of the cluster that local server can join in;
+// if the downgrade enabled status is true, the version window is [localVersion, oneMinorHigher]
+// if the downgrade is not enabled, the version window is [MinClusterVersion, localVersion]
+func decideAllowedVersionRange(downgradeEnabled bool) (minV *semver.Version, maxV *semver.Version) {
+	minV = semver.Must(semver.NewVersion(version.MinClusterVersion))
+	maxV = semver.Must(semver.NewVersion(version.Version))
+	maxV = &semver.Version{Major: maxV.Major, Minor: maxV.Minor}
+
+	if downgradeEnabled {
```

Comment:
This means enabling downgrade might reduce cluster availability. Considering a cluster of servers with version {4.1, 4.2, 4.2} which has cluster version of 4.1. Once enables downgrade to 4.0, a server of version 4.2 will not be able to restart if it crashes.

Can we retrieve downgrade target version from remote peers and use it as version guard?

Reply:
`allowedVersionRange` decides the cluster version range that the new member can join. Let's assume local version is 3.4 and `minClusterVersion` is 3.0. If downgrade is enabled, the new member can join in a cluster within [3.0, 3.5]. Otherwise, the range is [3.0, 3.4]
For restarting existing member, it is checked mostly at 
https://github.com/etcd-io/etcd/blob/984a89848ac63028fb5906fd6e6b69b254db4475/etcdserver/api/membership/cluster.go#L813:6